numpy/_core/src/multiarray/array_converter.c

```diff
@@ -0,0 +1,563 @@
+
+    self->narrs = 0;
+    self->flags = 0;
+    self->wrap = NULL;
+    self->wrap_type = NULL;
+
+    if (narrs == 0) {
+        return (PyObject *)self;
+    }
+    self->flags = (NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+
+    creation_item *item = self->items;
+    /* increase self->narrs in loop for cleanup */
+    for (int i = 0; i < narrs; i++, item++) {
+        item->object = PyTuple_GET_ITEM(args, i);
+
+        /* Fast path if input is an array (maybe FromAny should be faster): */
+        if (PyArray_Check(item->object)) {
+            Py_INCREF(item->object);
+            item->array = (PyArrayObject *)item->object;
+            item->scalar_input = 0;
+        }
+        else {
+            item->array = (PyArrayObject *)PyArray_FromAny_int(
+                    item->object, NULL, NULL, 0, 0, 0, NULL,
+                    &item->scalar_input);
+            if (item->array == NULL) {
+                goto fail;
+            }
+        }
+        if (!item->scalar_input) {
+            self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+        }
+        /* At this point, assume cleanup should happen for this item */
+        self->narrs++;
+        Py_INCREF(item->object);
+        item->DType = NPY_DTYPE(PyArray_DESCR(item->array));
+        Py_INCREF(item->DType);
+
+        if (npy_mark_tmp_array_if_pyscalar(
+                item->object, item->array, &item->DType)) {
+            /* just clear the flag again */
+            ((PyArrayObject_fields *)(item->array))->flags &= (
+                    ~NPY_ARRAY_WAS_PYTHON_LITERAL);
+            item->descr = NULL;
+        }
+        else {
+            self->flags &= ~NPY_CH_ALL_PYSCALARS;
+
+            item->descr = PyArray_DESCR(item->array);
+            Py_INCREF(item->descr);
+        }
+    }
+
+    return (PyObject *)self;
+
+  fail:
+    Py_DECREF(self);
+    return NULL;
+}
+
+
+static PyObject *
+array_converter_get_scalar_input(PyArrayArrayConverterObject *self)
+{
+    PyObject *ret = PyTuple_New(self->narrs);
+    if (ret == NULL) {
+        return NULL;
+    }
+
+    creation_item *item = self->items;
+    for (int i = 0; i < self->narrs; i++, item++) {
+        if (item->scalar_input) {
+            Py_INCREF(Py_True);
+            PyTuple_SET_ITEM(ret, i, Py_True);
+        }
+        else {
+            Py_INCREF(Py_False);
+            PyTuple_SET_ITEM(ret, i, Py_False);
+        }
+    }
+    return ret;
+}
+
+
+static int
+find_wrap(PyArrayArrayConverterObject *self)
+{
+    PyObject *wrap = NULL;
+    PyObject *wrap_type = NULL;
+    double priority = NPY_NAN;  /* silence compiler warnings */
+
+    if (self->wrap != NULL) {
+        return 0;  /* nothing to do */
+    }
+
+    for (int i = 0; i < self->narrs; i++) {
+        PyObject *obj = self->items[i].object;
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
+                Py_INCREF(Py_None);
```

Comment:
Strictly, you don't need to set the wrap here, given the `if (wrap == NULL)` clause at the end. (same for scalar)